src/XMonad/Core.hs

```diff
@@ -552,19 +567,149 @@ getXMonadDataDir :: X String
 getXMonadDataDir = asks (dataDir . directories)
 {-# DEPRECATED getXMonadDataDir "Use `asks (dataDir . directories)' instead." #-}
 
--- | Get the name of the file used to store the xmonad window state.
-stateFileName :: X FilePath
-stateFileName = (</> "xmonad.state") <$> getXMonadDataDir
+binFileName, buildDirName :: Directories -> FilePath
+binFileName  Directories{ cacheDir } = cacheDir </> "xmonad-" <> arch <> "-" <> os
+buildDirName Directories{ cacheDir } = cacheDir </> "build-" <> arch <> "-" <> os
 
--- | 'recompile force', recompile the xmonad configuration file when
--- any of the following apply:
---
---      * force is 'True'
---
---      * the xmonad executable does not exist
---
---      * the xmonad executable is older than xmonad.hs or any file in
---        the @lib@ directory (under the configuration directory).
+errFileName, stateFileName :: Directories -> FilePath
+errFileName   Directories{ dataDir } = dataDir </> "xmonad.errors"
+stateFileName Directories{ dataDir } = dataDir </> "xmonad.state"
+
+srcFileName, libFileName :: Directories -> FilePath
+srcFileName Directories{ cfgDir } = cfgDir </> "xmonad.hs"
+libFileName Directories{ cfgDir } = cfgDir </> "lib"
+
+buildScriptFileName, stackYamlFileName :: Directories -> FilePath
+buildScriptFileName Directories{ cfgDir } = cfgDir </> "build"
+stackYamlFileName   Directories{ cfgDir } = cfgDir </> "stack.yaml"
+
+-- | Compilation method for xmonad configuration.
+data Compile = CompileGhc | CompileStackGhc FilePath | CompileScript FilePath
+    deriving (Show)
+
+-- | Detect compilation method by looking for known file names in xmonad
+-- configuration directory.
+detectCompile :: Directories -> IO Compile
+detectCompile dirs = tryScript <|> tryStack <|> useGhc
+  where
+    buildScript = buildScriptFileName dirs
+    stackYaml = stackYamlFileName dirs
+
+    tryScript = do
+        guard =<< doesFileExist buildScript
+        isExe <- isExecutable buildScript
+        if isExe
+          then do
+            trace $ "XMonad will use build script at " <> show buildScript <> " to recompile."
+            pure $ CompileScript buildScript
+          else do
+            trace $ "XMonad will not use build script, because " <> show buildScript <> " is not executable."
+            trace $ "Suggested resolution to use it: chmod u+x " <> show buildScript
+            empty
+
+    tryStack = do
+        guard =<< doesFileExist stackYaml
+        canonStackYaml <- canonicalizePath stackYaml
```

Comment:
Isn't this a bit of an overkill? Won't a simple `makeAbsolute` suffice?

Reply:
I specifically wanted this to resolve symlinks so that people can place the `stack.yaml` file and the git clones elsewhere. Having sources and `.stack-work`s inside of `~/.config` feels wrong and some people won't like it.

Reply:
Oh alright then, that makes sense